File: fe/src/components/features/tasks/types/task.ts
export type TaskStatus = 'PENDING' | 'COMPLETED' | 'ARCHIVED';
export type RecurrencePattern = 'DAILY' | 'WEEKLY';
export type DayOfWeek = 'MONDAY' | 'TUESDAY' | 'WEDNESDAY' | 'THURSDAY' | 'FRIDAY' | 'SATURDAY' | 'SUNDAY';

export interface BaseTask {
    id: string;
    name: string;
    displayOrder: number;
    createdAt: string;
    updatedAt: string;
    status: TaskStatus;
    category?: string;
    type: 'ONE_TIME' | 'RECURRING';
}

export interface OneTimeTask extends BaseTask {
    type: 'ONE_TIME';
    dueDate: string;
    completedAt?: string;
}

export interface RecurringTask extends BaseTask {
    type: 'RECURRING';
    recurrencePattern: RecurrencePattern;
    dayOfWeek?: DayOfWeek;
    startDate: string;
    endDate?: string;
    nextDueDate: string;
    lastCompletedDate?: string;
}

export type Task = OneTimeTask | RecurringTask;

export interface CreateOneTimeTaskRequest {
    type: 'ONE_TIME';
    name: string;
    dueDate: string;
    category?: string;
}

export interface CreateRecurringTaskRequest {
    type: 'RECURRING';
    name: string;
    recurrencePattern: RecurrencePattern;
    dayOfWeek?: DayOfWeek;
    startDate: string;
    endDate?: string;
    category?: string;
}

export type CreateTaskRequest = CreateOneTimeTaskRequest | CreateRecurringTaskRequest;

export interface UpdateTaskRequest {
    name?: string;
    category?: string;
    dueDate?: string;
    recurrencePattern?: RecurrencePattern;
    dayOfWeek?: DayOfWeek;
    endDate?: string;
}

export interface CompleteTaskRequest {
    notes?: string;
}

export interface TaskOrderUpdate {
    id: string;
    displayOrder: number;
}

export interface ReorderTasksRequest {
    taskOrders: TaskOrderUpdate[];
}

export interface TaskCompletionResponse {
    id: string;
    completedDate: string;
    completedAt: string;
    notes?: string;
}

export interface ApiError {
    status: number;
    error: string;
    message: string;
    code: string;
    timestamp: string;
    errors?: Array<{
        field: string;
        message?: string;
    }>;
}

export interface TaskFilters {
    category?: string;
    status?: TaskStatus;
    dueDate?: string;
    startDate?: string;
    endDate?: string;
}

File: be/src/main/kotlin/ai/programujz/demo/domain/model/Task.kt
package ai.programujz.demo.domain.model

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate

sealed class Task {
    abstract val id: TaskId?
    abstract val name: String
    abstract val displayOrder: Int
    abstract val category: String?
    abstract val status: TaskStatus
    abstract val tags: List<Tag>
}

data class OneTimeTask(
    override val id: TaskId? = null,
    override val name: String,
    override val displayOrder: Int,
    override val category: String? = null,
    override val status: TaskStatus = TaskStatus.PENDING,
    override val tags: List<Tag> = emptyList(),
    val dueDate: LocalDate,
    val completedAt: Instant? = null
) : Task()

data class RecurringTask(
    override val id: TaskId? = null,
    override val name: String,
    override val displayOrder: Int,
    override val category: String? = null,
    override val status: TaskStatus = TaskStatus.PENDING,
    override val tags: List<Tag> = emptyList(),
    val recurrencePattern: RecurrencePattern,
    val dayOfWeek: DayOfWeek? = null,
    val startDate: LocalDate,
    val endDate: LocalDate? = null,
    val nextDueDate: LocalDate
) : Task()

enum class TaskStatus {
    PENDING,
    COMPLETED,
    ARCHIVED
}

enum class RecurrencePattern {
    DAILY,
    WEEKLY
}

data class TaskCompletion(
    val id: TaskCompletionId? = null,
    val taskId: TaskId,
    val completedDate: LocalDate,
    val completedAt: Instant = Instant.now(),
    val notes: String? = null
)